File: src/gh_forgejo_shim/codex_smoke.py
from __future__ import annotations

import os
import subprocess
import time
from dataclasses import dataclass
from typing import Callable, Mapping, Sequence


CODEX_PR_BOARD_FIELDS = (
    "additions,baseRefName,createdAt,deletions,headRefName,isDraft,"
    "number,state,title,updatedAt,url,mergeStateStatus,mergeable,statusCheckRollup"
)


@dataclass(frozen=True)
class SmokeProbe:
    name: str
    command: tuple[str, ...]
    optional: bool = False


@dataclass(frozen=True)
class SmokeResult:
    probe: SmokeProbe
    exit_code: int
    duration_ms: float
    stdout_bytes: int
    stderr_bytes: int
    stderr_excerpt: str = ""

    @property
    def ok(self) -> bool:
        return self.exit_code == 0 or self.probe.optional


Runner = Callable[..., subprocess.CompletedProcess[str]]


def codex_smoke_probes() -> tuple[SmokeProbe, ...]:
    return (
        SmokeProbe("gh version", ("gh", "--version")),
        SmokeProbe("gh auth status", ("gh", "auth", "status")),
        SmokeProbe("gh api user", ("gh", "api", "user")),
        SmokeProbe(
            "gh pr status",
            (
                "gh",
                "pr",
                "status",
                "--json",
                "number,title,url,headRefName,statusCheckRollup",
            ),
            optional=True,
        ),
        SmokeProbe(
            "gh pr list",
            (
                "gh",
                "pr",
                "list",
                "--state",
                "open",
                "--limit",
                "10",
                "--json",
                CODEX_PR_BOARD_FIELDS,
            ),
            optional=True,
        ),
        SmokeProbe("git repo root", ("git", "rev-parse", "--show-toplevel")),
        SmokeProbe("git remotes", ("git", "remote", "-v")),
        SmokeProbe(
            "git default branch",
            ("git", "symbolic-ref", "--quiet", "--short", "refs/remotes/origin/HEAD"),
            optional=True,
        ),
        SmokeProbe("git current branch", ("git", "branch", "--show-current"), optional=True),
        SmokeProbe(
            "git upstream",
            ("git", "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}"),
            optional=True,
        ),
        SmokeProbe(
            "git local and remote refs",
            (
                "git",
                "for-each-ref",
                "--format=%(refname:short)",
                "refs/heads",
                "refs/remotes",
            ),
            optional=True,
        ),
    )


def run_codex_smoke(
    *,
    cwd: str | None = None,
    env: Mapping[str, str] | None = None,
    runner: Runner = subprocess.run,
) -> list[SmokeResult]:
    values = dict(os.environ if env is None else env)
    results: list[SmokeResult] = []
    for probe in codex_smoke_probes():
        started = time.perf_counter()
        try:
            completed = runner(
                list(probe.command),
                cwd=cwd,
                env=values,
                check=False,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
            exit_code = int(completed.returncode)
            stdout = completed.stdout or ""
            stderr = completed.stderr or ""
        except OSError as exc:
            exit_code = 127
            stdout = ""
            stderr = str(exc)
        results.append(
            SmokeResult(
                probe=probe,
                exit_code=exit_code,
                duration_ms=(time.perf_counter() - started) * 1000,
                stdout_bytes=len(stdout.encode("utf-8")),
                stderr_bytes=len(stderr.encode("utf-8")),
                stderr_excerpt=_excerpt(stderr),
            )
        )
    return results


def format_codex_smoke(results: Sequence[SmokeResult]) -> str:
    lines = ["Codex probe smoke results:"]
    for result in results:
        marker = "ok" if result.exit_code == 0 else ("optional-fail" if result.probe.optional else "fail")
        lines.append(
            f"[{marker}] {result.probe.name}: exit {result.exit_code}, "
            f"{result.duration_ms:.0f} ms, stdout {result.stdout_bytes} B, stderr {result.stderr_bytes} B"
        )
        if result.stderr_excerpt:
            lines.append(f"      stderr: {result.stderr_excerpt}")
    required_failures = [result for result in results if result.exit_code != 0 and not result.probe.optional]
    if required_failures:
        lines.append(f"Required probe failures: {len(required_failures)}")
    else:
        lines.append("Required probes passed.")
    return "\n".join(lines)


def _excerpt(value: str, *, limit: int = 240) -> str:
    compact = " ".join(value.split())
    if len(compact) <= limit:
        return compact
    return compact[: limit - 1] + "…"

File: crates/gh-forgejo-shim/src/codex_smoke.rs
//! Codex probe smoke checks.

use std::collections::HashMap;
use std::path::Path;
use std::process::Command;
use std::time::Instant;

pub const CODEX_PR_BOARD_FIELDS: &str = "additions,baseRefName,createdAt,deletions,headRefName,isDraft,number,state,title,updatedAt,url,mergeStateStatus,mergeable,statusCheckRollup";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SmokeProbe {
    pub name: &'static str,
    pub command: Vec<&'static str>,
    pub optional: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SmokeResult {
    pub probe: SmokeProbe,
    pub exit_code: i32,
    pub duration_ms: f64,
    pub stdout_bytes: usize,
    pub stderr_bytes: usize,
    pub stderr_excerpt: String,
}

impl SmokeResult {
    pub fn ok(&self) -> bool {
        self.exit_code == 0 || self.probe.optional
    }
}

pub fn codex_smoke_probes() -> Vec<SmokeProbe> {
    vec![
        SmokeProbe {
            name: "gh version",
            command: vec!["gh", "--version"],
            optional: false,
        },
        SmokeProbe {
            name: "gh auth status",
            command: vec!["gh", "auth", "status"],
            optional: false,
        },
        SmokeProbe {
            name: "gh api user",
            command: vec!["gh", "api", "user"],
            optional: false,
        },
        SmokeProbe {
            name: "gh pr status",
            command: vec![
                "gh",
                "pr",
                "status",
                "--json",
                "number,title,url,headRefName,statusCheckRollup",
            ],
            optional: true,
        },
        SmokeProbe {
            name: "gh pr list",
            command: vec![
                "gh",
                "pr",
                "list",
                "--state",
                "open",
                "--limit",
                "10",
                "--json",
                CODEX_PR_BOARD_FIELDS,
            ],
            optional: true,
        },
        SmokeProbe {
            name: "git repo root",
            command: vec!["git", "rev-parse", "--show-toplevel"],
            optional: false,
        },
        SmokeProbe {
            name: "git remotes",
            command: vec!["git", "remote", "-v"],
            optional: false,
        },
        SmokeProbe {
            name: "git default branch",
            command: vec![
                "git",
                "symbolic-ref",
                "--quiet",
                "--short",
                "refs/remotes/origin/HEAD",
            ],
            optional: true,
        },
        SmokeProbe {
            name: "git current branch",
            command: vec!["git", "branch", "--show-current"],
            optional: true,
        },
        SmokeProbe {
            name: "git upstream",
            command: vec![
                "git",
                "rev-parse",
                "--abbrev-ref",
                "--symbolic-full-name",
                "@{u}",
            ],
            optional: true,
        },
        SmokeProbe {
            name: "git local and remote refs",
            command: vec![
                "git",
                "for-each-ref",
                "--format=%(refname:short)",
                "refs/heads",
                "refs/remotes",
            ],
            optional: true,
        },
    ]
}

pub fn run_codex_smoke(cwd: Option<&Path>, env: &HashMap<String, String>) -> Vec<SmokeResult> {
    codex_smoke_probes()
        .into_iter()
        .map(|probe| run_probe(probe, cwd, env))
        .collect()
}

pub fn format_codex_smoke(results: &[SmokeResult]) -> String {
    let mut lines = vec!["Codex probe smoke results:".to_string()];
    for result in results {
        let marker = if result.exit_code == 0 {
            "ok"
        } else if result.probe.optional {
            "optional-fail"
        } else {
            "fail"
        };
        lines.push(format!(
            "[{marker}] {}: exit {}, {:.0} ms, stdout {} B, stderr {} B",
            result.probe.name,
            result.exit_code,
            result.duration_ms,
            result.stdout_bytes,
            result.stderr_bytes
        ));
        if !result.stderr_excerpt.is_empty() {
            lines.push(format!("      stderr: {}", result.stderr_excerpt));
        }
    }

    let required_failures = results
        .iter()
        .filter(|result| result.exit_code != 0 && !result.probe.optional)
        .count();
    if required_failures == 0 {
        lines.push("Required probes passed.".to_string());
    } else {
        lines.push(format!("Required probe failures: {required_failures}"));
    }
    lines.join("\n")
}

pub fn required_failures(results: &[SmokeResult]) -> usize {
    results
        .iter()
        .filter(|result| result.exit_code != 0 && !result.probe.optional)
        .count()
}

fn run_probe(probe: SmokeProbe, cwd: Option<&Path>, env: &HashMap<String, String>) -> SmokeResult {
    let started = Instant::now();
    let mut command = Command::new(probe.command[0]);
    command.args(&probe.command[1..]).env_clear().envs(env);
    if let Some(path) = cwd {
        command.current_dir(path);
    }

    let (exit_code, stdout, stderr) = match command.output() {
        Ok(output) => (
            output.status.code().unwrap_or(1),
            output.stdout,
            output.stderr,
        ),
        Err(error) => (127, Vec::new(), error.to_string().into_bytes()),
    };

    SmokeResult {
        probe,
        exit_code,
        duration_ms: started.elapsed().as_secs_f64() * 1000.0,
        stdout_bytes: stdout.len(),
        stderr_bytes: stderr.len(),
        stderr_excerpt: excerpt(&String::from_utf8_lossy(&stderr)),
    }
}

fn excerpt(value: &str) -> String {
    let compact = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if compact.len() <= 240 {
        return compact;
    }
    format!("{}...", &compact[..237])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn probe_sequence_includes_codex_gh_and_git_checks() {
        let commands = codex_smoke_probes()
            .into_iter()
            .map(|probe| probe.command.join(" "))
            .collect::<Vec<_>>();

        assert!(commands.contains(&"gh --version".to_string()));
        assert!(commands.contains(&"gh auth status".to_string()));
        assert!(commands.contains(&"git rev-parse --show-toplevel".to_string()));
        assert!(
            commands.contains(&"git rev-parse --abbrev-ref --symbolic-full-name @{u}".to_string())
        );
    }

    #[test]
    fn formatter_marks_required_and_optional_failures() {
        let results = vec![
            SmokeResult {
                probe: SmokeProbe {
                    name: "gh version",
                    command: vec!["gh", "--version"],
                    optional: false,
                },
                exit_code: 127,
                duration_ms: 1.0,
                stdout_bytes: 0,
                stderr_bytes: 10,
                stderr_excerpt: "missing gh".to_string(),
            },
            SmokeResult {
                probe: SmokeProbe {
                    name: "gh pr status",
                    command: vec!["gh", "pr", "status"],
                    optional: true,
                },
                exit_code: 1,
                duration_ms: 1.0,
                stdout_bytes: 0,
                stderr_bytes: 10,
                stderr_excerpt: "no pull request".to_string(),
            },
        ];

        let output = format_codex_smoke(&results);

        assert!(output.contains("[fail] gh version"));
        assert!(output.contains("[optional-fail] gh pr status"));
        assert!(output.contains("Required probe failures: 1"));
        assert_eq!(required_failures(&results), 1);
    }
}
